Build failure: pkgsCross.armv7l-hf-multiplatform.gsoap

### Steps To Reproduce

(on 14d9c5542e569a9236011ec19ecd29c0009f0c19, latest master at the time of writing)
`nix-build -A pkgsCross.armv7l-hf-multiplatform.gsoap`

### Build log

```
[...]
Making all in wsdl
make[4]: Entering directory '/build/gsoap-2.8/gsoap/wsdl'
../../gsoap/src/soapcpp2 -SC -pwsdl -I../../gsoap/wsdl -I../../gsoap/import ../../gsoap/wsdl/wsdl.h
/nix/store/4bj2kxdm1462fzcc2i2s4dn33g2angcc-bash-5.2p32/bin/bash: line 1: ../../gsoap/src/soapcpp2: cannot execute binary file: Exec format error
make[4]: *** [Makefile:938: wsdlC.cpp] Error 126
make[4]: Leaving directory '/build/gsoap-2.8/gsoap/wsdl'
make[3]: *** [Makefile:830: all-recursive] Error 1
make[3]: Leaving directory '/build/gsoap-2.8/gsoap'
make[2]: *** [Makefile:458: all] Error 2
make[2]: Leaving directory '/build/gsoap-2.8/gsoap'
make[1]: *** [Makefile:475: all-recursive] Error 1
make[1]: Leaving directory '/build/gsoap-2.8'
make: *** [Makefile:383: all] Error 2
```

### Additional context

As stated in the title, I'm compiling for armhf. Looks like it's trying to execute an output binary as part of the build process.
I just want the libraries to link against in another project I'm packaging

Going back a couple (10k-50k) commits, this seems to have never worked, at least not recently

### Notify maintainers

@bjornfor 

<!--
Please @ people who are in the `meta.maintainers` list of the offending package or module.
If in doubt, check `git blame` for whoever last touched something.
-->

### Metadata

```console
➜ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.146.1-microsoft-standard-WSL2+, NixOS, 24.05 (Uakari), 24.05.20240822.797f7dc`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.5`
 - channels(root): `"nixos-23.11, nixos-wsl"`
 - nixpkgs: `/nix/store/5z81la6av4j0ckp0w1949lxiwjyyykks-source`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/development/libraries/gsoap/default.nix
+++ b/pkgs/development/libraries/gsoap/default.nix
@@ -1,7 +1,8 @@
-{ lib, stdenv, fetchurl, autoreconfHook, unzip, m4, bison, flex, openssl, zlib }:
+{ lib, stdenv, fetchurl, autoreconfHook, unzip, m4, bison, flex, openssl, zlib, buildPackages }:
 
 let
   majorVersion = "2.8";
+  isCross = stdenv.hostPlatform != stdenv.buildPlatform;
 
 in stdenv.mkDerivation rec {
   pname = "gsoap";
@@ -22,6 +23,10 @@ in stdenv.mkDerivation rec {
   prePatch = ''
     substituteInPlace configure.ac \
       --replace 'AM_INIT_AUTOMAKE([foreign])' 'AM_INIT_AUTOMAKE([foreign subdir-objects])'
+    ${lib.optionalString isCross ''
+      substituteInPlace gsoap/wsdl/Makefile.am \
+        --replace-fail 'SOAP=$(top_builddir)/gsoap/src/soapcpp2$(EXEEXT)' 'SOAP=${lib.getExe' buildPackages.gsoap "soapcpp2"}'
+    ''}
   '';
 
   meta = with lib; {

